### PR TITLE
feat: start test route implementation

### PR DESCRIPTION
**Pull request by petrSchreiber**
implements route for starting tests via final-api
- fixes few issues the current solution had

### CODE REVIEW

File: lib/final-api/endpoint/post/ddtf_tests/post_ddtf_tests.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            config:
+            {
+              language: 'tsd',
+              git:
+              {
+                no_clone: true
+              },
+              name: tsd['name'],
+              description: test_data.get_ikey('Description'),
+              branch: tsd['source']['git'] ? tsd['source']['git'] : tsd['source']['unc'],
```
Comment (ondrej-hosak) [inline lib/final-api/endpoint/post/ddtf_tests/post_ddtf_tests.rb]: this causes that branch is not correctly displayed on ATOM UI.

File: lib/final-api/endpoint/ddtf.rb [hunk 2]
```diff
@@ -1,114 +1,174 @@
...
+            request_body = request_body.merge('BuildId' => build.id)
+
+            cluster_name = enqueue_data.clusters.first
+            cluster_endpoint = FinalAPI.config.tsd_utils.clusters[cluster_name.to_sym]
+            guid = format('00000000-0000-0000-0000-%012d', build.id)
+
+            node_starter_data = {
+              build_id: build.id,
+              config: {
+                id: guid.to_s,
```
Comment (ondrej-hosak) [inline lib/final-api/endpoint/ddtf.rb]: no need to call to_s

File: lib/final-api/endpoint/ddtf.rb [hunk 3]
```diff
@@ -1,114 +1,176 @@
 require 'final-api/builder'
+require 'tsd_utils'
+
+require 'final-api/endpoint/post/ddtf_tests/post_ddtf_tests'
```
Comment (lksv) [inline lib/final-api/endpoint/ddtf.rb]: put helper file to lib/final-api/helpers/ddtf_tests_helper.rb

File: lib/final-api/endpoint/ddtf.rb [hunk 4]
```diff
@@ -1,114 +1,176 @@
...
+            halt 202
+          end
+
+          app.post '/ddtf/builds/:id/restart' do
+            service = Travis.service(:reset_model, current_user, build_id: params[:id])
+            halt(403, { messages: service.messages }.to_json) unless service.accept?
+            Travis.run_service(:reset_model, current_user, build_id: params['id'])
+            halt 202
+          end
+        end
```
Comment (lksv) [inline lib/final-api/endpoint/ddtf.rb]: private

... or better put it to helpers file.

File: lib/final-api/endpoint/ddtf.rb [hunk 5]
```diff
@@ -1,114 +1,176 @@
...
-          request = Request.create!(repository: repository, owner: User.find(params[:user_id]))
-
-          build = Build.create!(
-            repository: repository,
-            commit: commit,
-            request: request,
-            config: params[:config],
-            owner: request.owner
-          ) #creates job matrix, which I need to destroy
+        def postprocess_build(build)
```
Comment (lksv) [inline lib/final-api/endpoint/ddtf.rb]: rename please to `clean_build_matrix`

File: lib/final-api/endpoint/ddtf.rb [hunk 6]
```diff
@@ -1,114 +1,176 @@
...
           build.matrix.destroy_all
           build.cached_matrix_ids = nil
           build.start(started_at: Time.now.utc)
```
Comment (lksv) [inline lib/final-api/endpoint/ddtf.rb]: delete this line.

**Comment by lksv** [comment]
Use real user, repo and commit hash passed by params for 'POST /ddtf/test' if present (e.g. for uploaded tst use "uploadedTsd" project" and create a commit).

**Comment by lksv** [comment]
Please create issue for atom-ui to accept new response format for 'POST /ddtf/tests'

File: lib/final-api/endpoint/post/ddtf_tests/post_ddtf_tests.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
+require 'final-api/ext/hash'
+
+module FinalAPI::Endpoint
+  module DDTF
+    class PostDdtfTests
+      class << self
+        public
+        def get_new_build_params(test_data, tsd)
```
Comment (lksv) [inline lib/final-api/endpoint/post/ddtf_tests/post_ddtf_tests.rb]: rename please to `new_builds_params`. It should be connected with ddtf_helpers file.

**Comment by lksv** [comment]
if you parse request.body to json, name the var `payload`, not `request_body`. Request body is type of String.

**Comment by lksv** [comment]
'POST /ddtf/tests':
- 1.  ddtf_new_build_config_params (e.g. get_new_build_params) should return only config property.
- 2. obtain/create user, repo, commit and request (objects, not only id), by setup_build

That's mean, reduce DB requests to one time for each model.

**Comment by lksv** [comment]
test 'adds build id'. Try to load build model and test config prerties

**Comment by lksv** [comment]
Now it OK, please rebase it. It conflicts now with the master branch.

**Pull request merged**